tools/llzk-tblgen/OpCAPIGen.cpp

```diff
@@ -0,0 +1,825 @@
+  //  - MlirLocation location
+  //  - MlirOperationState state
+  //  - Operand/Attribute/Result parameters per `generateCAPIBuildParams()`
+  struct : GenStringFromOpPieces {
+    void genResultInferred(llvm::raw_ostream &os) override {
+      os << "  mlirOperationStateEnableResultTypeInference(&state);\n";
+    }
+    void genResult(
+        llvm::raw_ostream &os, const NamedTypeConstraint &result, const std::string &resultName
+    ) override {
+      if (result.isVariadic()) {
+        os << llvm::formatv(
+            "  mlirOperationStateAddResults(&state, {0}Size, {0}Types);\n", resultName
+        );
+      } else {
+        os << llvm::formatv("  mlirOperationStateAddResults(&state, 1, &{0}Type);\n", resultName);
+      }
+    }
+    void genOperand(llvm::raw_ostream &os, const NamedTypeConstraint &operand) override {
+      if (operand.isVariadic()) {
+        os << llvm::formatv(
+            "  mlirOperationStateAddOperands(&state, {0}Size, {0});\n", operand.name
+        );
+      } else {
+        os << llvm::formatv("  mlirOperationStateAddOperands(&state, 1, &{0});\n", operand.name);
+      }
+    }
+    void genAttributesPrefix(llvm::raw_ostream &os, const mlir::tblgen::Operator &op) override {
+      os << "  MlirContext ctx = mlirOpBuilderGetContext(builder);\n";
+      os << "  llvm::SmallVector<MlirNamedAttribute, " << op.getNumAttributes()
+         << "> attributes;\n";
+    }
+    void genAttribute(llvm::raw_ostream &os, const NamedAttribute &attr) override {
+      // The second parameter to `mlirNamedAttributeGet()` must be an "MlirAttribute". However, if
+      // it ends up as "MlirIdentifier", a reinterpret cast is needed. These C structs have the same
+      // layout and the C++ mlir::StringAttr is a subclass of mlir::Attribute so the cast is safe.
+      std::optional<std::string> attrType = tryCppTypeToCapiType(attr.attr.getStorageType());
+      std::string attrValue;
+      if (attrType.has_value() && attrType.value() == "MlirIdentifier") {
+        attrValue = "reinterpret_cast<MlirAttribute&>(" + attr.name.str() + ")";
+      } else {
+        attrValue = attr.name.str();
+      }
+
+      os << "  if (!mlirAttributeIsNull(" << attrValue << ")) {\n";
+      os << "    attributes.push_back(mlirNamedAttributeGet(mlirIdentifierGet(ctx, "
+         << "mlirStringRefCreateFromCString(\"" << attr.name << "\")), " << attrValue << "));\n";
+      os << "  }\n";
+    }
+    void genAttributesSuffix(llvm::raw_ostream &os, const mlir::tblgen::Operator &op) override {
+      os << "  mlirOperationStateAddAttributes(&state, attributes.size(), attributes.data());\n";
+    }
+    void genRegionsPrefix(llvm::raw_ostream &os, const mlir::tblgen::Operator &op) override {
+      os << "  llvm::SmallVector<MlirRegion, 2> regions;\n";
+      auto numRegions = op.getNumRegions();
+      if (numRegions > 2) { // pre-size if more than the initial capacity above
+        os << "  regions.resize(" << numRegions << ");\n";
+      }
```

Comment:
```suggestion
      os << "  llvm::SmallVector<MlirRegion, " << op.getNumRegions() << "> regions;\n";
```
No need to dynamically resize regions when we know the number of regions statically.